src/lib/core/prbuf.c

```diff
@@ -369,3 +376,166 @@ prbuf_iterator_next(struct prbuf_iterator *iter, struct prbuf_entry *result)
 	result->ptr = iter->current->data;
 	return 0;
 }
+
+void
+prbuf_reader_destroy(struct prbuf_reader *iter)
+{
+	ibuf_destroy(&iter->buf);
+}
+
+/**
+ * Check whether buffer header is sane.
+ */
+static int
+prbuf_check_header(struct prbuf_header *header)
+{
+	if (header->version > prbuf_version) {
+		diag_set(FileFormatError,
+			 "unknown format version %d", header->version);
+		return -1;
+	}
+
+	/* Check we can the read first record and end position is correct. */
+	if (prbuf_header_size + header->begin +
+				record_size_overhead >= header->size ||
+	    prbuf_header_size + header->end > header->size) {
+		diag_set(FileFormatError, "inconsistent header");
+		return -1;
+	}
+
+	return 0;
+}
+
+/** Read ahead is 128k. */
+#define PRBUF_READ_AHEAD (1 << 17)
+
+int
+prbuf_reader_create(struct prbuf_reader *reader, int fd, off_t offset)
+{
+	struct prbuf_header header;
+	/*
+	 * Read ahead does not make sense when reading prbuf header as
+	 * regularly first record is not at the beginning of the buffer.
+	 */
+	ssize_t rc = fio_pread(fd, &header, prbuf_header_size, offset);
```

Comment:
Let's read the header lazily so that the constructor never fails. The API will look simpler then.

Reply:
Done.